### DISCUSSION TITLE
CUDA Memory Requirements

### DISCUSSION THREAD

**Discussion by pman07** [toverainc/willow-inference-server#73] [Q&A] [answered]
Apologies if this is covered somewhere and I missed it, what do I need to change to get this running with 4 GB of VRAM? I've tried editing the tts_memory_threshold but I'm still running out of CUDA VRAM while "Warming models". Any tips? Thanks!

**Comment by kristiankielhofner** [comment]
What hardware are you using with 4GB VRAM?

**Comment by pman07** [comment]
I had an old GTX 1050 Ti laying around, was trying that out. Is that too low end?

**Reply by kristiankielhofner** [reply, marked as answer]
I just pushed a change to the wisng branch. I had to guess at the actual available reported CUDA memory but it should be close. I'm showing all of the Whisper models using 3.245GB of VRAM here in int8.

wisng will likely be merged to main this week so we'd really appreciate any testing you can do with it!

**Reply by pman07** [reply]
Cool, thanks for the quick response! I got wisng running after disabling TTS. Hopefully will get around to tinkering with it tomorrow.

**Reply by kristiankielhofner** [reply]
Great, let us know how it goes. The 1050 Ti is officially the lowest supported card anyone has ever run WIS on, interested to see how it performs.

**Reply by pman07** [reply]
Working pretty well, I think! I noticed the commit to revert to default willow response right about the time I was running into that issue. I'm seeing inference times of just under a second and still feels pretty quick for most uses so far. Thanks again for your awesome help! Is there anything you'd like me to test?

**Reply by kristiankielhofner** [reply]
Great! If you'd like to experiment with faster response times you can try one of the smaller models (such as small or base). A couple of options:

1) Copy settings.py to custom_settings.py and override the default model there.

2) Select the model on a per-request basis by adding the model param to the URI:

```/api/willow?model=small```

Theoretically the accuracy will be lower but for most voice assistant tasks, etc our default of medium probably leans a little heavy to the accuracy side - but I'd be very interested to hear what your experience is like! A couple of other general tips for reducing latency:

1) If it's acceptable for you and your environment you can use HTTP on port 19001 (by default) from Willow to WIS. This shaves some time off the initial connection to WIS on each session - we're looking at utilizing a persistent websocket connection from Willow to WIS to eliminate this. Additionally, we've identified a few performance issues with some of the TLS cipher suites on the ESP that we're investigating.

2) Speaking of which, you mentioned reverting to the default willow response - that should have only impacted you if you were using Willow + WIS (wisng) with Home Assistant pre 2023.5. If you use 2023.5 or later we maintain a persistent websocket connection to Home Assistant and in addition to being more capable it's also faster because of connection reuse.

3) If you speak relatively quickly (like I do) you can go to the Advanced Options section of the Willow Configuration menu and reduce your VAD timeout. The default of 300ms is also pretty conservative (I use 100ms).

I'll be adding formal benchmark support later this week and if you could run them on the 1050ti we'd like to add the results to our benchmarks.

**Reply by pman07** [reply]
I'll look into those performance tweaks, and I'd be more than happy to try running formal benchmarking.

The default willow response issue I was having was WIS returning a payload that included the inference time etc. which Home Assistant didn't like and gave a HTTP 400 response. The latest commit fixed that for me.

**Reply by kristiankielhofner** [reply]
Yes, that means you are using an a pre-2023.5 version of Home Assistant.

In versions before 2023.5 without Assist Pipeline support we fall back to the legacy conversation API. When I made that change I (wrongly) assumed that Home Assistant would accept JSON with extra fields but it does not. Anyway, I reverted that change.

With Home Assistant 2023.5 and onward we parse the JSON received from WIS and send it down a persistent websocket connection to HA. This has many advantages, the greatest of which is improved response time from speech to action (because we don't have to open a new connection) as well as improved flexibility and API compatibility moving forward.

In short, you should upgrade to HA 2023.5 whenever convenient, but I'm glad our multiple version support is working for you!

**Reply by pman07** [reply]
Sorry, I've been on the latest version of HA (2023.5.4), I think I worded my 'issue' poorly, but everything is working as expected!

**Reply by kristiankielhofner** [reply]
I think I'm understanding you.

What I keep coming back to is why the extra fields didn't work for you - when HomeAssistant 2023.5 is detected we use the Websockets Assist Pipeline API where we only parse the response text and pass it to HA via websockets - the extra fields (including language, unfortunately) are completely ignored. If your install was using websockets (as all of our dev/test environments do) you should not have been impacted at all by that commit. I only reverted it because in pre-2023.5 installs we pass the WIS response as-is to HomeAssistant via the legacy Conversation REST API.

Can you provide a full Willow boot log with at least one voice session to HA so we can look into this?

**Reply by pman07** [reply]
Well, I downloaded the latest version and reflashed. I believe it's working as intended now, see below for logs

> I (101) esp_image: segment 0: paddr=00010020 vaddr=3c1400I (166) esp_image: segment 1: paddr=0005a2b8 vaddr=3fc9f5d0 size=05d60h ( 23904) load
> I (171) esp_image: segment 2: paddr=00060020 vaddr=42000020 size=1388e8h (1280232) map
> I (401) esp_image: segment 3: paddr=00198910 vaddr=3fca5330 size=00c34h (  3124) load
> I (402) esp_image: segment 4: paddr=0019954c vaddr=40378000 size=175d0h ( 95696) load
> I (436) boot: Loaded app from partition at offset 0x10000
> I (437) boot: Disabling RNG early entropy source...
> I (448) opi psram: vendor id : 0x0d (AP)
> I (448) opi psram: dev id    : 0x02 (generation 3)
> I (449) opi psram: density   : 0x03 (64 Mbit)
> I (452) opi psram: good-die  : 0x01 (Pass)
> I (457) opi psram: Latency   : 0x01 (Fixed)
> I (462) opi psram: VCC       : 0x01 (3V)
> I (466) opi psram: SRF       : 0x01 (Fast Refresh)
> I (472) opi psram: BurstType : 0x01 (Hybrid Wrap)
> I (477) opi psram: BurstLen  : 0x01 (32 Byte)
> I (482) opi psram: Readlatency  : 0x02 (10 cycles@Fixed)
> I (488) opi psram: DriveStrength: 0x00 (1/1)
> I (494) spiram: Found 64MBit SPI RAM device
> I (498) spiram: SPI RAM mode: sram 80m
> I (502) spiram: PSRAM initialized, cache is in normal (1-core) mode.
> I (509) cpu_start: Pro cpu up.
> I (513) cpu_start: Starting app cpu, entry point is 0x403796c4
> I (0) cpu_start: App cpu up.
> I (802) spiram: SPI SRAM memory test OK
> I (810) cpu_start: Pro cpu start user code
> I (810) cpu_start: cpu freq: 240000000
> I (811) cpu_start: Application information:
> I (813) cpu_start: Project name:     willow
> I (818) cpu_start: App version:      1
> I (823) cpu_start: Compile time:     Jun  3 2023 05:19:10
> I (829) cpu_start: ELF file SHA256:  3d2207d674887a43...
> I (835) cpu_start: ESP-IDF:          v4.4.4-448-gfd5e03b221-dirty
> I (841) cpu_start: Min chip rev:     v0.0
> I (846) cpu_start: Max chip rev:     v0.99
> I (851) cpu_start: Chip rev:         v0.1
> I (856) heap_init: Initializing. RAM available for dynamic allocation:
> I (863) heap_init: At 3FCAADC8 len 0003E948 (250 KiB): D/IRAM
> I (869) heap_init: At 3FCE9710 len 00005724 (21 KiB): STACK/DIRAM
> I (876) heap_init: At 600FE000 len 00002000 (8 KiB): RTCRAM
> I (882) spiram: Adding pool of 8192K of external SPI memory to heap allocator
> I (891) spi_flash: detected chip: gd
> I (895) spi_flash: flash io: dio
> I (899) sleep: Configure to isolate all GPIO pins in sleep state
> I (905) sleep: Enable automatic switching of GPIO sleep configuration
> I (912) cpu_start: Starting scheduler on PRO CPU.
> I (0) cpu_start: Starting scheduler on APP CPU.
> I (928) spiram: Reserving pool of 16K of internal memory for DMA/internal allocations
> I (05:22:55.591) WILLOW: Starting up! Please wait...
> I (05:22:55.721) WILLOW: Waiting on SPIFFS mount...
> I (05:22:56.222) WILLOW: SPIFFS mounted
> W (05:23:00.084) PERIPH_WIFI: WiFi Event cb, Unhandle event_base:WIFI_EVENT, event_id:4
> I (05:23:01.585) WILLOW: initializing SNTP client
> I (11:23:01.587) WILLOW: Using configured SNTP server 'pool.ntp.org'
> I (11:23:01.613) WILLOW: audio_hal_ctrl_codec: ESP_OK
> I (11:23:01.614) WILLOW: HASS URL: http://192.168.1.82:8123/api/components
> I (11:23:01.634) WILLOW: HTTP status='200' content_length='4505'
> I (11:23:01.639) WILLOW: Home Assistant has Assist Pipeline support
> I (11:23:01.641) WILLOW: HASS URL: ws://192.168.1.82:8123/api/websocket
> I (11:23:01.653) WILLOW: WS event ID: 1
> E (11:23:01.658) lcd_panel.io.i2c: panel_io_i2c_rx_buffer(128): i2c transaction failed
> E (11:23:01.659) TT21100: esp_lcd_touch_tt21100_read_data(173): I2C read error!
> E (11:23:01.665) TT21100: esp_lcd_touch_new_i2c_tt21100(103): TT21100 init failed
> E (11:23:01.673) TT21100: Error (0xffffffff)! Touch controller TT21100 initialization failed!
> E (11:23:01.683) WILLOW: failed to initialize touch screen: ESP_FAIL
> I (11:23:01.691) WILLOW: audio_element_getinfo(hdl_ae_hs): sample_rate='44100' channels='2' bits='16' bps = '0'
> 
> ----------------------------- ESP Audio Platform -----------------------------
> |                                                                            |
> |                       ESP_AUDIO-v1.7.2-20e6bd0-b92a149                     |
> |                     Compile date: Nov 30 2022-07:50:12                     |
> ------------------------------------------------------------------------------
> I (11:23:01.739) WILLOW: audio player initialized
> E (11:23:01.741) I2S: register I2S object to platform failed
> I (11:23:01.749) WILLOW: Using record buffer '12'
> MC Quantized wakenet9: wakenet9l_v3h24_alexa_3_0.625_0.645, tigger:v3, mode:2, p:0, (May  5 2023 20:32:52)
> I (11:23:01.925) WILLOW: app_main() - start_rec() finished
> I (11:23:01.926) WILLOW: esp_netif_get_nr_of_ifs: 1
> I (11:23:01.928) WILLOW: MAC address: 7c:df:a1:e8:20:b8
> I (11:23:01.932) WILLOW: Startup complete! Waiting for wake word.
> I (11:23:03.429) WILLOW: SNTP client synchronized time to 1685769783
> I (11:23:12.604) WILLOW: Wake LCD timeout, turning off LCD
> I (11:23:19.186) WILLOW: AUDIO_REC_WAKEUP_START
> I (11:23:19.794) WILLOW: AUDIO_REC_VAD_START
> I (11:23:19.796) WILLOW: Using WIS URL 'http://192.168.1.84:19001/api/willow'
> I (11:23:19.797) WILLOW: WIS HTTP client starting stream, waiting for end of speech
> I (11:23:21.109) WILLOW: AUDIO_REC_VAD_END
> I (11:23:21.110) WILLOW: AUDIO_REC_WAKEUP_END
> I (11:23:21.136) WILLOW: WIS HTTP client HTTP_STREAM_POST_REQUEST, write end chunked marker
> I (11:23:22.135) WILLOW: WIS HTTP client HTTP_STREAM_FINISH_REQUEST
> I (11:23:22.135) WILLOW: WIS HTTP Response = {"language":"en","text":"Turn off computer room light."}
> I (11:23:22.142) WILLOW: sending command to Home Assistant via WebSocket: {
>         "end_stage":    "intent",
>         "id":   1685769802,
>         "input":        {
>                 "text": "Turn off computer room light."
>         },
>         "start_stage":  "intent",
>         "type": "assist_pipeline/run"
> }
> I (11:23:22.265) WILLOW: home assistant response_type: action_done
> I (11:23:22.270) WILLOW: received run-end event on WebSocket: {
>         "id":   1685769802,
>         "type": "event",
>         "event":        {
>                 "type": "run-end",
>                 "data": null,
>                 "timestamp":    "2023-06-03T05:23:22.285915+00:00"
>         }
> }
> I (11:23:33.283) WILLOW: Wake LCD timeout, turning off LCD
> I (11:24:56.715) WILLOW: AUDIO_REC_WAKEUP_START
> I (11:24:57.582) WILLOW: AUDIO_REC_VAD_START
> E (11:24:57.583) AUDIO_ELEMENT: [http_stream_writer] Element already stopped
> E (11:24:57.585) AUDIO_ELEMENT: [raw_stream_writer_to_api] Element already stopped
> I (11:24:57.595) WILLOW: Using WIS URL 'http://192.168.1.84:19001/api/willow'
> I (11:24:57.604) WILLOW: WIS HTTP client starting stream, waiting for end of speech
> I (11:24:58.642) WILLOW: AUDIO_REC_VAD_END
> I (11:24:58.643) WILLOW: AUDIO_REC_WAKEUP_END
> I (11:24:58.666) WILLOW: WIS HTTP client HTTP_STREAM_POST_REQUEST, write end chunked marker
> I (11:24:59.644) WILLOW: WIS HTTP client HTTP_STREAM_FINISH_REQUEST
> I (11:24:59.645) WILLOW: WIS HTTP Response = {"language":"en","text":"Turn on computer room light."}
> I (11:24:59.652) WILLOW: sending command to Home Assistant via WebSocket: {
>         "end_stage":    "intent",
>         "id":   1685769899,
>         "input":        {
>                 "text": "Turn on computer room light."
>         },
>         "start_stage":  "intent",
>         "type": "assist_pipeline/run"
> }
> I (11:24:59.777) WILLOW: home assistant response_type: action_done
> I (11:24:59.783) WILLOW: received run-end event on WebSocket: {
>         "id":   1685769899,
>         "type": "event",
>         "event":        {
>                 "type": "run-end",
>                 "data": null,
>                 "timestamp":    "2023-06-03T05:24:59.796163+00:00"
>         }
> }
> I (11:25:10.887) WILLOW: Wake LCD timeout, turning off LCD

**Reply by kristiankielhofner** [reply]
> WILLOW: sending command to Home Assistant via WebSocket:

Yep, looks good!

**Comment by ceinstaller** [comment]
pman, can you list the settings you used for your 1050ti?  I'm using the same card but can't seem to find the right combination of settings to get the server to start.  Always hit the 'CUDA failed with error out of memory'.

Thanks in advance,

~Mark

**Reply by pman07** [reply]
@ceinstaller Sorry for the delay, I was out of town. Are you using the 'wisng' branch? I don't believe I had to do anything special to get that branch running with the 1050ti.

**Reply by ceinstaller** [reply]
Hey pman, I now owe you an apology for a late reply.  I didn't get it to work, but I was using the instructions from the main page.  (I assume that means I wasn't using the wsing branch.)  I ended up pulling a 1060 out of another machine and using that and all is well!
